server/src/main/java/io/crate/role/TransportPrivilegesAction.java

```diff
@@ -91,61 +91,70 @@ protected void masterOperation(PrivilegesRequest request,
 
     }
 
     @VisibleForTesting
-    static List<String> validateUserNames(Metadata metadata, Collection<String> userNames) {
+    static List<String> validateRoleNames(Metadata metadata, PrivilegesRequest request) {
         RolesMetadata rolesMetadata = metadata.custom(RolesMetadata.TYPE);
+        Set<String> roleNamesToCheck = new LinkedHashSet<>(request.userNames()); // Keep order for exception message
+        if (request.rolePrivilege() != null) {
+            roleNamesToCheck.addAll(request.rolePrivilege().roleNames());
+        }
         if (rolesMetadata == null) {
-            return new ArrayList<>(userNames);
+            return new ArrayList<>(roleNamesToCheck);
         }
-        List<String> unknownUserNames = null;
-        for (String userName : userNames) {
+        List<String> unknownRoleNames = new ArrayList<>();
+        for (String roleName : roleNamesToCheck) {
             //noinspection PointlessBooleanExpression
-            if (rolesMetadata.roleNames().contains(userName) == false) {
-                if (unknownUserNames == null) {
-                    unknownUserNames = new ArrayList<>();
-                }
-                unknownUserNames.add(userName);
+            if (rolesMetadata.roleNames().contains(roleName) == false) {
+                unknownRoleNames.add(roleName);
             }
         }
-        if (unknownUserNames == null) {
-            return Collections.emptyList();
-        }
-        return unknownUserNames;
+        return unknownRoleNames;
     }
 
     @VisibleForTesting
     static long applyPrivileges(Metadata.Builder mdBuilder,
                                 PrivilegesRequest request) {
-        // create a new instance of the metadata, to guarantee the cluster changed action.
-        UsersPrivilegesMetadata newMetadata = UsersPrivilegesMetadata.copyOf(
-            (UsersPrivilegesMetadata) mdBuilder.getCustom(UsersPrivilegesMetadata.TYPE));
-
-        long affectedRows = newMetadata.applyPrivileges(request.userNames(), request.privileges());
-        mdBuilder.putCustom(UsersPrivilegesMetadata.TYPE, newMetadata);
-        return affectedRows;
+        long affectedRows;
+        if (request.privileges().isEmpty() == false) {
+            // create a new instance of the metadata, to guarantee the cluster changed action.
+            UsersPrivilegesMetadata newMetadata = UsersPrivilegesMetadata.copyOf(
+                (UsersPrivilegesMetadata) mdBuilder.getCustom(UsersPrivilegesMetadata.TYPE));
+
+            affectedRows = newMetadata.applyPrivileges(request.userNames(), request.privileges());
+            mdBuilder.putCustom(UsersPrivilegesMetadata.TYPE, newMetadata);
+            return affectedRows;
+        } else {
+            // create a new instance of the metadata, to guarantee the cluster changed action.
+            RolesMetadata newMetadata = RolesMetadata.newInstance(
+                (RolesMetadata) mdBuilder.getCustom(UsersPrivilegesMetadata.TYPE));
+
```

Comment:
Why `UsersPrivilegesMetadata.TYPE` is used here? Shouldn't be `RolesMetadata.TYPE`?

Reply:
yep, sorry, something wrong when cleanup up, I'll push it together with the integ test.